### DISCUSSION TITLE
[Restructured in Wiki] How should OSmosis caps be sent between PDs?

### DISCUSSION THREAD

**Discussion by p-linh** [sid-agrawal/OSmosis#17] [General]
We have a `pd_client_send_cap` API call that does exactly what you think it does:
**RPD** = Receiver PD, **SPD** = Sender PD

1. If the sent cap was unwrapped into a badge, mint a new badged endpoint for RPD, and place this endpoint into RPD's CSpace. SPD will need to make RPD aware of what slot this endpoint is in.
2. If the sent cap is a regular cap, just `CNode_Copy` it into RPD's CSpace.

This worked fine previously when we did not carry any additional info in the GPI context structs (e.g. the `pd_client_context_t`s) and the client did not need to know about this additional info. However, the system has evolved enough such that the client *does* need to have additional info, sometimes it being an object ID or something (for various reasons). 

If we use `pd_client_send_cap`, the only thing RPD receives is the endpoint in some slot and cannot retrieve this extra info unless the resource server can place that info somewhere (which requires them to have a shared message buffer) or it makes an additional call to the resource server to ask for that info, based on the given badged endpoint. 

Additionally, we weren't using `pd_client_send_cap` for anything other than sending regular, unbadged endpoints to other PDs. I had to slightly modify it so that we could send required caps (e.g. the ADS, CPU) to a child PD and have the RT update this info in the PD's OSmosis data frame.  We can't do this for *all* types of resources though, and so this begs the question of how we should be sending GPI-specific resources to other PDs?

- Do we keep it as is, and if RPD needs the additional info, it must request it from the server? Or SPD has to somehow also give RPD this info?
     - I don't like this because it means the data is fragmented and has to be retrieved in two different ways (e.g. reading from a slot + info from another PD)
- Use a portion of the OSmosis data frame every PD has to just place received GPI resources
     - I also don't like this because RPD would have to read the resource out of the data frame immediately after it's sent, otherwise it risks being overwritten by another GPI cap send
- Build with the fact that RPD will not know about additional info from a received cap in its slot - and try to limit the additional info as much as possible

---
Aside as to why this problem came up: We want to refcount regular endpoints, and we track them as if they were GPI resources (just for the sake of reusing code) but they are not canonical resources under the framework. However, this requires that we include both the GPI badged endpoint *and* the raw, unbadged, regular endpoint in a resource context representing the regular endpoint. When we try to send this endpoint to another PD, we need to somehow send both the GPI badged one and the regular raw one (or decide that we only need to send one of them).

**Comment by astevins** [comment]
Is it possible that the endpoint context is *just* the badged endpoint? Then if a PD wants to receive the endpoint itself into its own cspace, it can make a call to the RT.

**Reply by p-linh** [reply]
I was thinking that when we send endpoints, RPD only gets the regular unbadged one, and without a context containing the badged RT endpoint, it can't send it to other PDs (we also will have to disallow sending of non-GPI caps through the PD component for this), which seems reasonable to me. 

Also with resources in general, if we have additional info, do we want to just always want to have to make an additional call to some resource server upon receiving a badged endpoint?

**Reply by astevins** [reply]
Are we sure that we should limit the RPD so it cannot perform any resource operations on the endpoint? That doesn't make sense to me.

**Reply by p-linh** [reply]
What other resource operations would it need to perform on the badged endpoint? It only needs to listen on the regular endpoint? For destroying the endpoint, I would think that it shouldn't be the one allowed to invoke a "destroy" on it?

**Reply by astevins** [reply]
It seems that you are using send_cap to send the endpoint to the PD that will listen on it and do nothing else. But I don't see why we should restrict it to this exact usage. What if the PD wants to delegate the fault handling to some other PD? What if it wants to free the endpoint to indicate that it won't be handling faults anymore? Etc.

**Reply by p-linh** [reply]
If PD1 allocated the EP, and wants to delegate handling to PD2, I'm thinking that PD2 just needs to listen on the endpoint and nothing else. In the case where PD2 wants to indicate it won't be handling faults anymore - I would assume that for our system, we can just simplify this into the case where PD2 stops executing and dies (and the recount of the EP gets reduced). 

Although I can't really see a use-case for it now, maybe restricting this will cause other issues in the future.
